src/darwin/Framework/CHIP/MTRDelegateManager.h

```diff
@@ -0,0 +1,89 @@
+/**
+ *    Copyright (c) 2025 Project CHIP Authors
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+#import "Foundation/Foundation.h"
+
+#import "MTRDefines_Internal.h"
+
+NS_ASSUME_NONNULL_BEGIN
+
+/**
+ * MTRDelegateInfo can be subclassed to store additional information, in
+ * addition to the delegate and queue.
+ */
+MTR_TESTABLE
+@interface MTRDelegateInfo<__covariant DelegateType> : NSObject
+- (instancetype)initWithDelegate:(DelegateType)delegate queue:(dispatch_queue_t)queue;
+
+@property (nonatomic, weak, readonly) DelegateType delegate;
+@property (nonatomic, readonly) dispatch_queue_t queue;
+
+// Pointer value for logging purpose only
+@property (readonly) void * delegatePointerValue;
+
+// Queues an async call to the provided block on the delegate's queue if the
+// delegate still exists.  Returns whether the delegate exists.
+- (BOOL)callDelegateWithBlock:(void (^)(id delegate))block;
+
+@end
+
+/**
+ * A delegate manager that handles automatic removal of delegates that have gone
+ * away and provides helpers for working with delegates.  This interface does
+ * not do any synchronization; access to an MTRDelegateManager should be
+ * synchronized by its consumer as needed.
+ *
+ * It sure would be nice if the type bound on our second type parameter could be
+ * MTRDelegateInfo<DelegateType>, but that does not seem to be supported.
+ */
+MTR_TESTABLE
+@interface MTRDelegateManager<DelegateType, DelegateInfoType : MTRDelegateInfo *> : NSObject
+
+- (instancetype)initWithOwner:(id)owner;
```

Comment:
So the owner is just so we can include it in log messages?

Reply:
Seems to me a convenient way to do it this way, especially if the owner has description that changes.

Reply:
Correct.  I could call it "ownerForLogging" if desired?

Reply:
I think leaving it as owner is fine, maybe just put a comment.

Reply:
Done, and documented that a weak ref is held to owner.